src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,170 @@ def pileupSizeTask(self):
+        :param fraction: container fraction
+        :return: list of blocks
+        """
+        pname = doc['pileupName']
+        lastTransition = doc['transition'][-1]
+        cname = lastTransition['customDID']
+        totalBlocks = self.rucioClient.getBlocksInContainer(pname)
+        customBlocks = self.rucioClient.getBlocksInContainer(cname)
+        portion = math.ceil(fraction * len(totalBlocks))
+        blockList = customBlocks + [b for b in totalBlocks if b not in customBlocks]
+        self.logger.info("increase scenario: use %d blocks out of %d custom blocks from %s and %d from %s",
+                         len(blockList), len(customBlocks), cname, len(totalBlocks), pname)
+        return blockList[:portion]
+
+    def getDecreasingBlocks(self, doc, fraction):
+        """
+        Return list of blocks for decrease scenarios of partial pileup
+        according to logic outlined in
+        https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5#logic-for-increasingdecreasing-container-fraction
+
+        :param doc: pileup document
+        :param fraction: container fraction
+        :return: list of blocks
+        """
+        pname = doc['pileupName']
+        cname = doc['customName']
+        totalBlocks = self.rucioClient.getBlocksInContainer(pname)
+        portion = math.ceil(fraction * len(totalBlocks))
+        if cname == '':
+            self.logger.info("decrease scenario: use %d blocks out of pileup %s with %d blocks",
+                             len(totalBlocks[:portion]), pname, len(totalBlocks))
+            return totalBlocks[:portion]
+        customBlocks = self.rucioClient.getBlocksInContainer(cname)
+        self.logger.info("decrease scenario: use %d blocks out of custom container %s",
+                         len(customBlocks[:portion]), cname)
+        return customBlocks[:portion]
+
+    def partialPileupTask(self):
+        """
+        Execute partial pileup placement according to the following logic:
+
+        - get rucio DIDs (datasets, in CMS name blocks) for our pileup document
+        - get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+        - create new container DID in Rucio via Rucio wrapper createContainer API
+          which calls rucio CLI add_container
+        - update MSPileup JSON with
+          - create new custom Name as pileup+extention
+          - add new transition record and update MSPileup document
+        - we call attachDIDs Rucio wrapper API with our set of DIDs and rses from pileup document
+        - create new rules for custom DID via provided rseExpression
+        - add new ruleIds to pileup document
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+
+        :return: None
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
```

Comment:
Just wondering if this isn't too verbose? Perhaps we could have it as a debug level, and have a shorter info level log like "partialPileupTask processing pileup name %s", doc['pileupName']"

Reply:
ok